eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala

```diff
@@ -458,6 +458,19 @@ case class Commitment(fundingTxIndex: Long,
     } else if (missingForReceiver < 0.msat) {
       if (params.localParams.isInitiator) {
         // receiver is not the channel initiator; it is ok if it can't maintain its channel_reserve for now, as long as its balance is increasing, which is the case if it is receiving a payment
+      } else if (reduced.toLocal > fees && reduced.htlcs.size < 5) {
```

Comment:
After a dual funded channel creation or splice the reserve should ideally be met for both sides - but if it's not required when the new funding tx is created, then consistency with single funded channels would imply that "[the protocol ensures that there is always progress toward meeting this reserve, and once met, it is maintained.](https://github.com/lightning/bolts/blob/4dcc377209509b13cf89a4b91fde7d478f5b46d8/02-peer-protocol.md?plain=1#L296)"

One way to interpret this is that if the receiver's previous commitment would have been below the required reserve with the current fee rate and capacity, we should also allow this update because it will make progress towards increasing their reserve.  Something like this:

```scala
    val missingForReceiverBefore = remoteCommit1.spec.toLocal - remoteChannelReserve(params)- (if (params.localParams.isInitiator) 0.sat else fees)
    if (missingForSender < 0.msat) {
      return Left(InsufficientFunds(params.channelId, amount = amount, missing = -missingForSender.truncateToSatoshi, reserve = localChannelReserve(params), fees = if (params.localParams.isInitiator) fees else 0.sat))
    } else if (missingForReceiver < 0.msat) {
      if (params.localParams.isInitiator) {
        // receiver is not the channel initiator; it is ok if it can't maintain its channel_reserve for now, as long as its balance is increasing, which is the case if it is receiving a payment
      } else if (reduced.toLocal > fees && missingForReceiverBefore < 0.msat) {
```
However, once in the "previous was below reserve" state the receiver/initiator could maliciously cause the sender to route many htlcs through them until their reserve is zero after paying the commitment tx fees.  The receiver/initiator could then fail those htlcs and later try to publish the revoked commitment. The receiver/initiator risks losing commit tx fees equivalent to their previous reserve. The sender would only be able to reclaim the htlc value, but not the reserve amount, if they catch the cheating attempt.

Even if the malicous receiver/initiator is a miner, they would still lose the opportunity to make those fees from another transaction if the commitment tx fee rates are competitive. At best, it seems like more of a griefing attack against the sender than a way for the receiver/initiator to make a risk-free attack.

The proposed solution of allowing only 5 pending htlcs at a time should prevent this possible attack. As long as those 5 htlcs are not failed, the receiver/initiator's reserve will increase and the channel will not become stuck. Once the receiver/initiator has a balance that exceeds the new reserve and fee buffer, they can route payments back to the non-initiator.

My questions are:
 - Have I understood the situation correctly?
 - Would a sender that just paid on-chain fees to spliced-in value to the channel rather take the of risk a complicated griefing attack over the risk of losing routing fees by failing htlcs that come in faster than they are resolved? Especially if failing them could reduce the attractiveness of a profitable route?
 - Is it possible to compute if the previous remote commitment would have been below the reserve requirements based on our current capacity and fee rates? This might be easier than trying to track the last reserve limit.

Reply:
> One way to interpret this is that if the receiver's previous commitment would have been below the required reserve with the current fee rate and capacity, we should also allow this update because it will make progress towards increasing their reserve.

I'm not sure you can simply compare to `remoteCommit1`, since there may be many changes in-flight I'm afraid this will miss some subtle edge cases.

Also this won't work in the following edge case (and probably other subtle similar edge cases):

- the previous commitment is slightly below its reserve
- we make a splice than slightly _decreases_ the reserve
- the previous commitment is slightly above the new reserve
- but the current commitment is below the new reserve
- we still want to allow some HTLCs to unblock that situation

> As long as those 5 htlcs are not failed, the receiver/initiator's reserve will increase and the channel will not become stuck.

I believe it's even slightly better than that: if HTLCs are fulfilled, we're making progress, if they're failed, we're simply back to where we were before.

> Have I understood the situation correctly?

I believe you did, depending on whether my last comment makes sense for you or not!

> Would a sender that just paid on-chain fees to spliced-in value to the channel rather take the of risk a complicated griefing attack over the risk of losing routing fees by failing htlcs that come in faster than they are resolved? 

I don't understand this scenario. By `sender` do you mean the sender of the HTLC? The potential attacker here isn't the node who spliced funds in, but rather the node who can use this opportunity to dip into their reserve (if they were the channel initiator).

But game-theory-wise, the node that could potentially be under attack just spliced a lot of funds into the channel, providing free inbound liquidity to the potential attacker: I believe this is a good incentive to not perform the attack and just benefit from free inbound liquidity!

> Is it possible to compute if the previous remote commitment would have been below the reserve requirements based on our current capacity and fee rates? This might be easier than trying to track the last reserve limit.

See the beginning of this comment, I believe this isn't possible, or is really subtle and error-prone. I'd rather have a simpler solution that we can easily reason about. The risk with this solution is very simple to analyze: can the increased weight from 5 additional HTLCs dip low enough into the channel reserve to create an incentive to cheat? Since anchor outputs, the commit tx feerate is kept to a low value, so the increased fee for 5 HTLC outputs is negligible, which is why I think this solution is acceptable.

Reply:
> > Would a sender that just paid on-chain fees to spliced-in value to the channel rather take the of risk a complicated griefing attack over the risk of losing routing fees by failing htlcs that come in faster than they are resolved?
> 
> I don't understand this scenario. By `sender` do you mean the sender of the HTLC? The potential attacker here isn't the node who spliced funds in, but rather the node who can use this opportunity to dip into their reserve (if they were the channel initiator).
> 
> But game-theory-wise, the node that could potentially be under attack just spliced a lot of funds into the channel, providing free inbound liquidity to the potential attacker: I believe this is a good incentive to not perform the attack and just benefit from free inbound liquidity!

A simpler way to put my question: is 5 in-flight htlcs too few? how about 10? 50?

The scenario I'm thinking about above is related to your second point that the game-theory for the attacker is not strong. There is a risk that valid fee paying htlcs will be rejected when we put any sort of limitation on the htlcs that the `non-initiator`/`sender` might send to a potentially attacking `initiator`/`receiver` that is below their reserve through no fault of their own. This is a mostly a question that looks at the problem the other way: in practice is `5` in-flight htlcs too few? especially if the feerate is low due to anchor outputs.

Because there is also a cost to rejecting htlcs in both fees lost and potentially route reputation, how can we quantify the maximum we accept? Do we have any empirical data on how many in-flight htlcs a busy channel typically has at any given time?

Reply:
> > One way to interpret this is that if the receiver's previous commitment would have been below the required reserve with the current fee rate and capacity, we should also allow this update because it will make progress towards increasing their reserve.
> 
> I'm not sure you can simply compare to `remoteCommit1`, since there may be many changes in-flight I'm afraid this will miss some subtle edge cases.

Right - I agree, better to have a simple limit than risk difficult to predict edge cases.

Reply:
> A simpler way to put my question: is 5 in-flight htlcs too few? how about 10? 50?

That's a very good question. One additional HTLC in the commitment transaction costs 43 sats in additional fees at 1 sat/byte. Thus 5 HTLCs at 20 sat/byte cost 4300 sats in additional fees (and you simpy need to multiply to get the numbers for X HTLC at Y sat/byte).

What we ideally want is that this extra cost doesn't lead to the initiator's main output being trimmed (so that they always have *something* to lose if they publish a revoked commitment). Most implementations reject channels that are smaller than 100 000 sat, for which the reserve will thus be 1000 sat.

For such channels, the initiator's main output would end up being trimmed very quickly. But for such a small amount, is it really an issue? The non-initiator doesn't lose any money if the initiator publishes a revoked commitment, it's simply that the initiator doesn't get penalized (apart from the on-chain fees they pay, which means we're in the same model at Eltoo).

What I want to convey here is that even 1 additional HTLC can end up trimming the initiator's main output (if the channel is small enough or the commit feerate high enough). We don't *need* to allow many HTLCs, so I believe we should just allow enough that they are likely to be fulfilled. I chose 5 instead of 1 because of MPP: allowing only 1 HTLC has the risk that it is part of an MPP for which we reject the other parts, which will prevent the HTLC from being fulfilled. But 5 is completely arbitrary and I don't think there is a good, scientific way of choosing that number...